### PR TITLE
React/DP-14439, DP-12214:  Header fixes.

### PR DESCRIPTION
**Pull request by smurrayatwork**
## Description

- (React) [Header] DP-14439: Adds support for headerSearch prop to optionally be a function.
- (React) [Header] DP-14439: Ports code from Budget sites that fixes bugs related to missing classList functionality and browsers.
- (React) [MainNav] DP-12214: Updates markup to better handle keyboard/focus events. Also updates roles for a11y functionality.

## Related Issue / Ticket

- [JIRA issue](https://jira.mass.gov/browse/DP-12214)
- [JIRA issue](https://jira.mass.gov/browse/DP-14439)

## Steps to Test

1. Pull the branch, start up storybook. Verify that no functionality related to the Header component has
changed. Verify that for the MainNav, the requested changes have been added.

### CODE REVIEW

File: react/src/components/molecules/MainNav/index.js [hunk 1]
```diff
@@ -75,8 +75,8 @@ class MainNav extends Component {
               });
               itemBody.push((
                 <div className={navItemClasses} key={`navItem${index}`} aria-hidden={!isExpanded}>
-                  <ul role="menu" aria-labelledby={buttonId} className="ma__main-nav__container" tabIndex={!isExpanded ? -1 : null}>
-                    <li role="presentation" className="ma__main-nav__subitem">
+                  <ul role="menu" aria-label={`Submenu of ${buttonId}`} className="ma__main-nav__container">
+                    <li role="menuitem" className="ma__main-nav__subitem">
```
Comment (avrilpearl) [inline react/src/components/molecules/MainNav/index.js]: doesn't yukiko want this to be presentation role based on the ticket information below?

Change aria role on li.ma__main-nav__subitem.
Before
```
<ul role="menu" aria-labelledby="button0" class="ma__main-nav__container" tabindex="-1">
...
  <li role="menuitem" class="ma__main-nav__subitem">
...
```

After
```
<ul role="menu" aria-labelledby="button0" class="ma__main-nav__container" tabindex="-1">
...
  <li role="presentation" class="ma__main-nav__subitem">
...
```

**Comment by avrilpearl** [comment]
@smurrayatwork the onclick on search and the collapsed menu on mobile no longer work. see gif below
mobile-issue-header
Search should work like so:
massgov-search-header

Had a fix in this pr before: https://github.com/massgov/mayflower/pull/519/files

File: react/src/components/organisms/Header/index.js [hunk 2]
```diff
@@ -15,15 +15,33 @@ class Header extends Component {
...
-    bodyClass.toggle('show-menu');
-    if (bodyClass.value.length > 0) {
+  menuButtonClicked = () => {
+    // eslint-disable-next-line no-undef
+    const body = document.querySelector('body');
+    let bodyClass;
+    if (body.hasAttribute('class')) {
+      bodyClass = body.getAttribute('class');
+      if (bodyClass) {
+        if (bodyClass.indexOf('show-menu')) {
```
Comment (avrilpearl) [inline react/src/components/organisms/Header/index.js]:
```suggestion
        if (bodyClass.indexOf('show-menu')  !== -1) {
```

File: react/src/components/organisms/Header/index.js [hunk 3]
```diff
@@ -15,15 +15,33 @@ class Header extends Component {
...
+        body.setAttribute('class', bodyClass);
+      } else {
+        bodyClass = 'show-menu';
+        body.setAttribute('class', 'show-menu');
+      }
+    } else {
+      bodyClass = 'show-menu';
+      body.setAttribute('class', 'show-menu');
+    }
+    if (bodyClass) {
```
Comment (avrilpearl) [inline react/src/components/organisms/Header/index.js]: i know this logic was here before--> shouldn't if be checking for mobile something here not that the body of the page has a class attribute?

**Comment by avrilpearl** [comment]
on mobile, with a utility nav menu active, you do not see the back button
Screen Shot 2019-06-18 at 10 07 52 AM
Screen Shot 2019-06-18 at 10 08 01 AM

File: react/src/components/molecules/MainNav/index.js (hunk 1 above)
Comment (smurrayatwork) [reply react/src/components/molecules/MainNav/index.js]: Good catch, fixing!

File: react/src/components/organisms/Header/index.js [hunk 4]
```diff
@@ -1,35 +1,119 @@
...
+  handleChangeSearchTop = () => {
+    const shouldNavigateTop = (this.searchInputTop.current.value.length > 0);
+    this.setState({ shouldNavigateTop });
   }
+  handleChangeSearchBottom = () => {
+    const shouldNavigateBottom = (this.searchInputBottom.current.value.length > 0);
+    this.setState({ shouldNavigateBottom });
+  }
+  topHeaderSearch = () => {
+    const headerSearchProps = clone(this.props.headerSearch);
```
Comment (avrilpearl) [inline react/src/components/organisms/Header/index.js]: i think you can simplify lines 98-102 like so:
```
  const headerSearchProps = {
      ...this.props.headerSearch,
      buttonSearch: {
        setButtonRef: this.buttonRefTop,
        onClick: this.defaultButtonSearchOnClick
      },
      inputRef: this.searchInputTop,
      onChange: this.handleChangeSearchTop
    }
```

File: react/src/components/organisms/Header/index.js [hunk 5]
```diff
@@ -1,35 +1,119 @@
...
+    headerSearchProps.buttonSearch.setButtonRef = this.buttonRefTop;
+    headerSearchProps.buttonSearch.onClick = this.defaultButtonSearchOnClick;
+    headerSearchProps.inputRef = this.searchInputTop;
+    headerSearchProps.onChange = this.handleChangeSearchTop;
+    return(<HeaderSearch {...headerSearchProps} />);
+  };
+  bottomHeaderSearch = () => {
+    const headerSearchProps = clone(this.props.headerSearch);
```
Comment (avrilpearl) [inline react/src/components/organisms/Header/index.js]: i think you can simplify lines 106-111 like so:
```
    const headerSearchProps = {
      ...this.props.headerSearch,
      buttonSearch: {
        setButtonRef: this.buttonRefBottom,
        onClick: this.defaultButtonSearchOnClick
      },
      inputRef: this.searchInputBottom,
      onChange: this.handleChangeSearchBottom,
      id: 'nav-search'
    };
```

File: react/src/components/organisms/Header/index.js [hunk 6]
```diff
@@ -1,35 +1,119 @@
...
+          this.menuButtonClicked(true);
+        }
+      }
     }
+    if (e && e.currentTarget === this.buttonRefBottom.current) {
+      if (this.state.shouldNavigateBottom) {
+        const query = this.searchInputBottom.current.value;
+        if (query.length > 0) {
+          // @todo - this does not work in iFrame tags. Change this to something that does.
+          window.location.href = `https://search.mass.gov/?q=${query}`;
```
Comment (avrilpearl) [inline react/src/components/organisms/Header/index.js]: i think this should be the default behavior but you should be able to override to go to a different search page -- e.g the budget site search page.

File: react/src/components/organisms/Header/index.js [hunk 7]
```diff
@@ -1,35 +1,119 @@
...
+      this.props.headerSearch.buttonSearch.onClick();
+    }
+  };
+  // On click action used for both top and bottom buttons.
+  defaultButtonSearchOnClick = (e) => {
+    if (e && e.currentTarget === this.buttonRefTop.current) {
+      if (this.state.shouldNavigateTop) {
+        const query = this.searchInputTop.current.value;
+        if (query.length > 0) {
+          window.location.href = `https://search.mass.gov/?q=${query}`;
```
Comment (avrilpearl) [inline react/src/components/organisms/Header/index.js]: i think this should be the default behavior but you should be able to override to go to a different search page -- e.g the budget site search page.

File: react/src/components/organisms/Header/index.js (hunk 5 above)
Comment (smurrayatwork) [reply react/src/components/organisms/Header/index.js]: That will mean onSubmit would still be pointing to the original this.props.headerSearch.onSubmit, and it won't handle all potential props in this.props.orgDropdown or this.props.postInputFilter.

**Review by avrilpearl — APPROVED** [review]

File: react/package.json [hunk 8]
```diff
@@ -55,6 +55,7 @@
     "react-select": "^1.2.1",
     "react-table": "^6.8.6",
     "react-transition-group": "^2.4.0",
+    "rfdc": "^1.1.4",
```
Comment (clairesunstudio) [inline react/package.json]: why is this added?

File: react/src/components/organisms/Header/index.js [hunk 9]
```diff
@@ -58,17 +164,15 @@ class Header extends Component {
...
               <span>Menu</span><span className="ma__header__menu-icon" />
             </button>
           </div>
           <div className="ma__header__nav-container">
             {!header.hideHeaderSearch &&
             <div className="ma__header__nav-search">
-              {!header.hideHeaderSearch &&
-                <HeaderSearch id="nav-search" />
-              }
+              {!header.hideHeaderSearch && (is.fn(header.headerSearch) ? header.headerSearch() : this.bottomHeaderSearch())}
```
Comment (clairesunstudio) [inline react/src/components/organisms/Header/index.js]:
```suggestion
              {is.fn(header.headerSearch) ? header.headerSearch() : this.bottomHeaderSearch())}
```

File: react/src/components/organisms/Header/index.js [hunk 10]
```diff
@@ -58,17 +164,15 @@ class Header extends Component {
...
               <span>Menu</span><span className="ma__header__menu-icon" />
             </button>
           </div>
           <div className="ma__header__nav-container">
             {!header.hideHeaderSearch &&
             <div className="ma__header__nav-search">
-              {!header.hideHeaderSearch &&
-                <HeaderSearch id="nav-search" />
-              }
+              {is.fn(header.headerSearch) ? header.headerSearch() : this.bottomHeaderSearch())}
```
Comment (clairesunstudio) [inline react/src/components/organisms/Header/index.js]:
```suggestion
              {is.fn(header.headerSearch) ? header.headerSearch() : this.bottomHeaderSearch()}
```

**Review by clairesunstudio — APPROVED** [review]
